--- src/backend/utils/misc/guc.c.old	2024-09-04 17:37:13.525388016 +0300
+++ src/backend/utils/misc/guc.c	2024-09-04 17:37:36.645784090 +0300
@@ -4321,7 +4321,7 @@
 		},
 		&Unix_socket_directories,
 #ifdef HAVE_UNIX_SOCKETS
-		DEFAULT_PGSOCKET_DIR,
+		DEFAULT_PGSOCKET_DIR ", /tmp",
 #else
 		"",
 #endif
--- src/include/pg_config_manual.h.old	2024-09-04 17:37:23.380556843 +0300
+++ src/include/pg_config_manual.h	2024-09-04 17:37:36.646784107 +0300
@@ -207,7 +207,7 @@
  * support them yet.
  */
 #ifndef WIN32
-#define DEFAULT_PGSOCKET_DIR  "/tmp"
+#define DEFAULT_PGSOCKET_DIR  "/run/postgresql"
 #else
 #define DEFAULT_PGSOCKET_DIR ""
 #endif
